artifact.go

```diff
@@ -216,39 +465,193 @@ func (m ArtifactModule) postBuildArtifactHandler(c *gin.Context) {
 				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
 					"Failed with starting to read file content from multipart form request body when uploading new artifact for build with ID %d.",
 					buildID))
-				return
+				return nil, false
 			}
 
 			data, err := ioutil.ReadAll(f)
 			if err != nil {
 				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
 					"Failed reading file content from multipart form request body when uploading new artifact for build with ID %d.",
 					buildID))
-				return
+				return nil, false
 			}
 
-			artifact := Artifact{
-				Data:     data,
-				Name:     k,
-				FileName: fhs[0].Filename,
-				BuildID:  buildID,
-			}
+			files = append(files, &file{
+				name:     k,
+				fileName: fhs[0].Filename,
+				data:     data,
+			})
```

Comment:
IMO excess use of pointers. Suggest making the `file.data` field an enforced pointer instead, so you don't have to spray pointer dereferencing and such everywhere

Reply:
From some googling it seems like the slice's underlying array isn't copied when passing the slice by value. With this in mind I have changed it so that all the things are no longer pointers.

Not sure if that's the way to go, but the code looks cleaner now.

I also made sure to initialize the slices with a length to avoid unnecessary allocations.

Reply:
I always get a little confused by Go's difference between slice vs array. But if I understand it correctly, you're always working with slices unless you specify the size inside the brackets. And only if you specify the brackets, then the copy operation is expensive.

```go
var myArray = [3]int{1, 2, 3}

var mySlice = []int{1, 2, 3}
var myOtherSlice = make([]int, 3, 3)

var myArrayCopied = myArray // expensive op on big arrays
var myArrayAsSlice = myArray[:] // cheap op, just taking a ref

type myType struct {
    sliceField []int
    arrayField [3]int
}

ex1 := myType{sliceField: myArray}        // ok
ex2 := myType{sliceField: mySlice}        // ok
ex3 := myType{sliceField: make([]int, 3)} // ok

ex4 := myType{arrayField: myArray}        // ok, but expensive op on big arrays
ex5 := myType{arrayField: mySlice}        // compile error
ex6 := myType{arrayField: make([]int, 3)} // compile error
}
```

As you say, bottom line is that we're only working with slices anyway, and I think trying to use arrays here would be overkill when slices clearly suffices, even if we force the GC to deal with them as there's not 10 000 items in them

Reply:
Seems in line with what I read and understood 👍